cli: TestRemoveDeadReplicas failed

[(cli).TestRemoveDeadReplicas failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2062941&tab=buildLog) on [master@456a07cfc1e53b87abc7709052e54efb1450e758](https://github.com/cockroachdb/cockroach/commits/456a07cfc1e53b87abc7709052e54efb1450e758):

```
=== RUN   TestRemoveDeadReplicas/2/4/r=4
W200704 06:07:10.112919 176174 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
W200704 06:07:10.220959 176569 kv/kvserver/store.go:1656  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
W200704 06:07:10.440045 176174 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
W200704 06:07:10.463437 176174 gossip/gossip.go:1504  [n?] no incoming or outgoing connections
W200704 06:07:10.510884 176174 jobs/registry.go:561  [n2] unable to get node liveness: node not in the liveness table
W200704 06:07:10.514814 176174 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
W200704 06:07:10.533440 176174 gossip/gossip.go:1504  [n?] no incoming or outgoing connections
W200704 06:07:10.587817 176174 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
W200704 06:07:10.596631 176174 gossip/gossip.go:1504  [n?] no incoming or outgoing connections
W200704 06:07:13.757583 269291 kv/kvserver/intentresolver/intent_resolver.go:440  failed to push during intent resolution: failed to push meta={id=be7c138a key=/Local/Range/Table/14/RangeDescriptor pri=0.00224742 epo=0 ts=1593842833.743310055,0 min=1593842833.743310055,0 seq=7} lock=true stat=PENDING rts=0,0 wto=false max=0,0


ERROR: signal: segmentation fault (core dumped)

68 runs completed, 1 failures, over 49m19s
context canceled

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestRemoveDeadReplicas PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRemoveDeadReplicas.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3224890&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3224890&tab=artifacts#/) on master @ [64c888ffa869dc254f6f14dbbcb308efc1d89a64](https://github.com/cockroachdb/cockroach/commits/64c888ffa869dc254f6f14dbbcb308efc1d89a64):


```
Replica r30:/Table/3{4-5} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6]
Replica r31:/Table/3{5-6} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r31:/Table/3{5-6} [(n2,s2):5, next=6, gen=6]
Replica r32:/Table/3{6-7} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r32:/Table/3{6-7} [(n2,s2):5, next=6, gen=6]
Replica r33:/Table/3{7-8} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6]
Replica r34:/Table/3{8-9} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6]
Replica r35:/Table/{39-40} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6]
Replica r36:/Table/4{0-1} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6]
Replica r37:/Table/4{1-2} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6]
Replica r38:/Table/4{2-3} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6]
Replica r39:/Table/4{3-4} [(n2,s2):4, (n4,s4):2, (n3,s3):3, (n1,s1):5, next=6, gen=10] -> r39:/Table/4{3-4} [(n2,s2):6, next=7, gen=10]
Replica r40:/{Table/44-Max} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r40:/{Table/44-Max} [(n2,s2):5, next=6, gen=6]
Conflicting intent found on /Local/Range/System/tsd/RangeDescriptor. Aborting txn 4b6c143c-0486-449e-8c80-0330d28fa85a to resolve.
Scanning replicas on store cluster_id:89b23597-8ca7-4ca4-b31a-921720a4fb55 node_id:2 store_id:2  for dead peers []
E210726 13:52:55.232404 602306 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 49  leaseholder replica not in descriptor; desc: r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.674256304,0
E210726 13:52:55.232736 602306 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 50  leaseholder replica not in descriptor; desc: r32:/Table/3{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.615328380,0
E210726 13:52:55.277725 602137 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 51  leaseholder replica not in descriptor; desc: r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.484308102,0
E210726 13:52:55.306658 603341 kv/kvserver/replica.go:894  [n2,s2,r6/5:/Table/{SystemCon…-11}] 52  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.520998611,0
E210726 13:52:55.307060 603341 kv/kvserver/replica.go:894  [n2,s2,r6/5:/Table/{SystemCon…-11}] 53  leaseholder replica not in descriptor; desc: r7:/Table/1{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.564228537,0
E210726 13:52:55.320256 603270 kv/kvserver/replica.go:894  [n2,s2,r35/5:/Table/{39-40}] 54  leaseholder replica not in descriptor; desc: r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.636176881,0
E210726 13:52:55.351444 603270 kv/kvserver/replica.go:894  [n2,s2,r35/5:/Table/{39-40}] 57  leaseholder replica not in descriptor; desc: r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.519866286,0
E210726 13:52:55.348396 603360 kv/kvserver/replica.go:894  [n2,s2,r27/5:/{NamespaceTab…-Table/32}] 56  leaseholder replica not in descriptor; desc: r28:/Table/3{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.610938719,0
E210726 13:52:55.327544 603363 kv/kvserver/replica.go:894  [n2,s2,r26/5:/NamespaceTable/{30-Max}] 55  leaseholder replica not in descriptor; desc: r25:/{Table/29-NamespaceTable/30} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.651377682,0
E210726 13:52:55.406003 603598 kv/kvserver/replica.go:894  [n2,s2,r7/5:/Table/1{1-2}] 58  leaseholder replica not in descriptor; desc: r8:/Table/1{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.596936639,0
E210726 13:52:55.427376 603385 kv/kvserver/replica.go:894  [n2,s2,r4/5:/System{/tsd-tse}] 59  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.520998611,0
E210726 13:52:55.432339 604172 kv/kvserver/replica.go:894  [n2,s2,r11/6:/Table/1{5-6}] 60  leaseholder replica not in descriptor; desc: r12:/Table/1{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n2,s2):3 seq=2 start=1627307554.706612488,0 epo=1 pro=1627307554.706216845,0
E210726 13:53:00.416150 603328 kv/kvserver/replica_range_lease.go:365  [n1,s1,r16/1:/Table/2{0-1}] 61  failed to heartbeat own liveness record: context canceled

id,is_live,replicas,is_decommissioning,membership,is_draining
3,false,40,true,decommissioning,false
4,false,40,true,decommissioning,false
E210726 13:53:01.536331 652395 kv/kvserver/replica.go:894  [n2,s2,r14/5:/Table/1{8-9}] 62  leaseholder replica not in descriptor; desc: r13:/Table/1{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.643970560,0
E210726 13:53:01.680627 371747 kv/kvserver/replica.go:894  [n2,s2,r14/5:/Table/1{8-9}] 63  leaseholder replica not in descriptor; desc: r15:/Table/{19-20} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627307553.585342772,0 epo=1 pro=1627307553.585208363,0
E210726 13:53:02.646825 673531 kv/kvserver/replica.go:894  [n2,s2,r23/5:/Table/2{7-8}] 64  leaseholder replica not in descriptor; desc: r22:/Table/2{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.619848565,0
E210726 13:53:02.647000 673531 kv/kvserver/replica.go:894  [n2,s2,r23/5:/Table/2{7-8}] 65  leaseholder replica not in descriptor; desc: r24:/Table/2{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.665469459,0
E210726 13:53:03.194984 684410 kv/kvserver/replica.go:894  [n2,s2,r30/5:/Table/3{4-5}] 66  leaseholder replica not in descriptor; desc: r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.641213740,0
E210726 13:53:03.396618 688504 kv/kvserver/replica.go:894  [n2,s2,r28/5:/Table/3{2-3}] 67  leaseholder replica not in descriptor; desc: r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.641213740,0
E210726 13:53:03.613227 686698 kv/kvserver/replica.go:894  [n2,s2,r17/5:/Table/2{1-2}] 68  leaseholder replica not in descriptor; desc: r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.696127949,0
E210726 13:53:03.855008 697565 kv/kvserver/replica.go:894  [n2,s2,r38/5:/Table/4{2-3}] 69  leaseholder replica not in descriptor; desc: r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.691954397,0
E210726 13:53:04.060630 371747 kv/kvserver/replica.go:894  [n2,s2,r38/5:/Table/4{2-3}] 70  leaseholder replica not in descriptor; desc: r39:/Table/4{3-4} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):3 seq=2 start=1627307559.054348943,0 epo=1 pro=1627307559.054213879,0
E210726 13:53:04.431285 371747 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 71  leaseholder replica not in descriptor; desc: r15:/Table/{19-20} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627307553.585342772,0 epo=1 pro=1627307553.585208363,0
E210726 13:53:04.985999 720935 kv/kvserver/replica.go:894  [n2,s2,r21/5:/Table/2{5-6}] 72  leaseholder replica not in descriptor; desc: r20:/Table/2{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.677864568,0
E210726 13:53:04.986970 720935 kv/kvserver/replica.go:894  [n2,s2,r21/5:/Table/2{5-6}] 73  leaseholder replica not in descriptor; desc: r22:/Table/2{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.619848565,0
E210726 13:53:05.200904 726162 kv/kvserver/replica.go:894  [n2,s2,r36/5:/Table/4{0-1}] 74  leaseholder replica not in descriptor; desc: r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.691954397,0
E210726 13:53:05.940500 732454 kv/kvserver/replica.go:894  [n2,s2,r20/5:/Table/2{4-5}] 75  leaseholder replica not in descriptor; desc: r19:/Table/2{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.605058185,0
E210726 13:53:06.854163 751479 kv/kvserver/replica.go:894  [n2,s2,r25/5:/{Table/29-NamespaceTab…}] 76  leaseholder replica not in descriptor; desc: r24:/Table/2{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.665469459,0
E210726 13:53:07.727031 767845 kv/kvserver/replica.go:894  [n2,s2,r32/5:/Table/3{6-7}] 77  leaseholder replica not in descriptor; desc: r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.589716696,0
E210726 13:53:08.252637 779472 kv/kvserver/replica.go:894  [n2,s2,r33/5:/Table/3{7-8}] 78  leaseholder replica not in descriptor; desc: r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.636176881,0
E210726 13:53:09.444539 803404 kv/kvserver/replica.go:894  [n2,s2,r19/5:/Table/2{3-4}] 79  leaseholder replica not in descriptor; desc: r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627307552.696127949,0
    debug_test.go:399: expected replicas on {1,2,5,6} but got {1,2,5,7}
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (43.21s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRemoveDeadReplicas PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/server @cockroachdb/kv bdarnell
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3225210&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3225210&tab=artifacts#/) on master @ [64c888ffa869dc254f6f14dbbcb308efc1d89a64](https://github.com/cockroachdb/cockroach/commits/64c888ffa869dc254f6f14dbbcb308efc1d89a64):


```
Replica r35:/Table/{39-40} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6]
Replica r36:/Table/4{0-1} [(n3,s3):4, (n2,s2):2, (n4,s4):3, (n1,s1):5, next=6, gen=10] -> r36:/Table/4{0-1} [(n2,s2):6, next=7, gen=10]
Replica r37:/Table/4{1-2} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6]
Replica r38:/Table/4{2-3} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6]
Replica r39:/Table/4{3-4} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r39:/Table/4{3-4} [(n2,s2):5, next=6, gen=6]
Replica r40:/{Table/44-Max} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r40:/{Table/44-Max} [(n2,s2):5, next=6, gen=6]
Conflicting intent found on /Local/Range/System/tsd/RangeDescriptor. Aborting txn c2d5cd0a-f19e-448f-807c-b515c604222f to resolve.
Scanning replicas on store cluster_id:e980aa9e-4144-4296-ab2e-e98dca22dab3 node_id:2 store_id:2  for dead peers []
E210726 15:04:51.397265 606653 kv/kvserver/replica.go:894  [n2,s2,r2/5:/System/NodeLiveness{-Max}] 44  leaseholder replica not in descriptor; desc: r3:/System/{NodeLivenessMax-tsd} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.606657911,0
E210726 15:04:51.513359 607675 kv/kvserver/replica.go:894  [n2,s2,r25/5:/{Table/29-NamespaceTab…}] 45  leaseholder replica not in descriptor; desc: r24:/Table/2{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.743504859,0
E210726 15:04:51.534223 608196 kv/kvserver/replica.go:894  [n2,s2,r34/5:/Table/3{8-9}] 46  leaseholder replica not in descriptor; desc: r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.727161999,0
E210726 15:04:51.535189 608196 kv/kvserver/replica.go:894  [n2,s2,r34/5:/Table/3{8-9}] 47  leaseholder replica not in descriptor; desc: r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.622601431,0
E210726 15:04:51.536064 607676 kv/kvserver/replica.go:894  [n2,s2,r6/5:/Table/{SystemCon…-11}] 48  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.696044065,0
E210726 15:04:51.536214 607676 kv/kvserver/replica.go:894  [n2,s2,r6/5:/Table/{SystemCon…-11}] 49  leaseholder replica not in descriptor; desc: r7:/Table/1{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.635868889,0
E210726 15:04:51.571494 608696 kv/kvserver/replica.go:894  [n2,s2,r3/5:/System/{NodeLive…-tsd}] 50  leaseholder replica not in descriptor; desc: r4:/System{/tsd-tse} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.643613662,0
E210726 15:04:51.592118 608835 kv/kvserver/replica.go:894  [n2,s2,r27/5:/{NamespaceTab…-Table/32}] 51  leaseholder replica not in descriptor; desc: r28:/Table/3{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.843558054,0
E210726 15:04:51.594184 608784 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 52  leaseholder replica not in descriptor; desc: r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.689845080,0
E210726 15:04:51.594354 608784 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 53  leaseholder replica not in descriptor; desc: r32:/Table/3{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.673113663,0
E210726 15:04:51.596864 608883 kv/kvserver/replica.go:894  [n2,s2,r7/5:/Table/1{1-2}] 54  leaseholder replica not in descriptor; desc: r8:/Table/1{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.652820706,0
E210726 15:04:51.616782 608854 kv/kvserver/replica.go:894  [n2,s2,r4/5:/System{/tsd-tse}] 55  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.696044065,0
E210726 15:04:51.624051 609154 kv/kvserver/replica.go:894  [n2,s2,r8/5:/Table/1{2-3}] 56  leaseholder replica not in descriptor; desc: r9:/Table/1{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.662457956,0
E210726 15:04:51.625621 609202 kv/kvserver/replica.go:894  [n2,s2,r11/5:/Table/1{5-6}] 57  leaseholder replica not in descriptor; desc: r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.600133038,0
E210726 15:04:51.710121 609924 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 58  leaseholder replica not in descriptor; desc: r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.873504599,0
E210726 15:04:54.957826 607652 kv/kvserver/replica_range_lease.go:365  [n1,s1,r3/1:/System/{NodeLive…-tsd}] 59  failed to heartbeat own liveness record: heartbeat failed on epoch increment
E210726 15:04:54.961866 607724 kv/kvserver/replica_range_lease.go:365  [n1,s1,r9/1:/Table/1{3-4}] 60  failed to heartbeat own liveness record: heartbeat failed on epoch increment
E210726 15:04:55.037225 633027 kv/kvserver/replica.go:894  [n2,s2,r9/5:/Table/1{3-4}] 61  leaseholder replica not in descriptor; desc: r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.600133038,0

id,is_live,replicas,is_decommissioning,membership,is_draining
3,false,40,true,decommissioning,false
4,false,40,true,decommissioning,false
E210726 15:04:57.057224 649509 kv/kvserver/replica.go:894  [n2,s2,r39/5:/Table/4{3-4}] 62  leaseholder replica not in descriptor; desc: r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.716999513,0
E210726 15:04:57.057420 649509 kv/kvserver/replica.go:894  [n2,s2,r39/5:/Table/4{3-4}] 63  leaseholder replica not in descriptor; desc: r40:/{Table/44-Max} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.698368724,0
E210726 15:04:57.332216 653142 kv/kvserver/replica.go:894  [n2,s2,r38/5:/Table/4{2-3}] 64  leaseholder replica not in descriptor; desc: r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.852409374,0
E210726 15:04:58.200520 669474 kv/kvserver/replica.go:894  [n2,s2,r24/5:/Table/2{8-9}] 65  leaseholder replica not in descriptor; desc: r23:/Table/2{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.733641927,0
E210726 15:04:58.635166 363864 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 66  leaseholder replica not in descriptor; desc: r15:/Table/{19-20} [(n2,s2):6, next=7, gen=10], lease: repl=(n4,s4):2 seq=2 start=1627311872.327294142,0 epo=1 pro=1627311872.327188788,0
E210726 15:04:58.684748 679748 kv/kvserver/replica.go:894  [n2,s2,r28/5:/Table/3{2-3}] 67  leaseholder replica not in descriptor; desc: r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.759058214,0
E210726 15:04:59.220138 688234 kv/kvserver/replica.go:894  [n2,s2,r30/5:/Table/3{4-5}] 68  leaseholder replica not in descriptor; desc: r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.759058214,0
E210726 15:04:59.937054 363864 kv/kvserver/replica.go:894  [n2,s2,r11/5:/Table/1{5-6}] 69  leaseholder replica not in descriptor; desc: r12:/Table/1{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):3 seq=2 start=1627311875.874064482,0 epo=1 pro=1627311875.873868278,0
E210726 15:05:00.297552 712796 kv/kvserver/replica.go:894  [n2,s2,r21/5:/Table/2{5-6}] 70  leaseholder replica not in descriptor; desc: r20:/Table/2{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.838563662,0
E210726 15:05:00.297800 712796 kv/kvserver/replica.go:894  [n2,s2,r21/5:/Table/2{5-6}] 71  leaseholder replica not in descriptor; desc: r22:/Table/2{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.859496437,0
E210726 15:05:00.944876 725740 kv/kvserver/replica.go:894  [n2,s2,r19/5:/Table/2{3-4}] 72  leaseholder replica not in descriptor; desc: r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.722770041,0
E210726 15:05:00.945143 725740 kv/kvserver/replica.go:894  [n2,s2,r19/5:/Table/2{3-4}] 73  leaseholder replica not in descriptor; desc: r20:/Table/2{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.838563662,0
E210726 15:05:01.175872 729641 kv/kvserver/replica.go:894  [n2,s2,r13/6:/Table/1{7-8}] 74  leaseholder replica not in descriptor; desc: r12:/Table/1{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):3 seq=2 start=1627311875.874064482,0 epo=1 pro=1627311875.873868278,0
E210726 15:05:01.176115 729641 kv/kvserver/replica.go:894  [n2,s2,r13/6:/Table/1{7-8}] 75  leaseholder replica not in descriptor; desc: r14:/Table/1{8-9} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627311873.427853766,0 epo=1 pro=1627311873.427716955,0
E210726 15:05:02.280460 746640 kv/kvserver/replica.go:894  [n2,s2,r17/5:/Table/2{1-2}] 76  leaseholder replica not in descriptor; desc: r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.722770041,0
E210726 15:05:02.805996 757156 kv/kvserver/replica.go:894  [n2,s2,r15/6:/Table/{19-20}] 77  leaseholder replica not in descriptor; desc: r14:/Table/1{8-9} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627311873.427853766,0 epo=1 pro=1627311873.427716955,0
E210726 15:05:04.308027 783619 kv/kvserver/replica.go:894  [n2,s2,r32/5:/Table/3{6-7}] 78  leaseholder replica not in descriptor; desc: r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.727161999,0
E210726 15:05:05.371157 804511 kv/kvserver/replica.go:894  [n2,s2,r22/5:/Table/2{6-7}] 79  leaseholder replica not in descriptor; desc: r23:/Table/2{7-8} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627311867.733641927,0
    debug_test.go:399: expected replicas on {1,2,5,6} but got {1,2,6,7}
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (43.54s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRemoveDeadReplicas PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/server @cockroachdb/kv bdarnell
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3225683&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3225683&tab=artifacts#/) on master @ [64c888ffa869dc254f6f14dbbcb308efc1d89a64](https://github.com/cockroachdb/cockroach/commits/64c888ffa869dc254f6f14dbbcb308efc1d89a64):


```
Replica r35:/Table/{39-40} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6]
Replica r36:/Table/4{0-1} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6]
Replica r37:/Table/4{1-2} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6]
Replica r38:/Table/4{2-3} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6]
Replica r39:/Table/4{3-4} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r39:/Table/4{3-4} [(n2,s2):5, next=6, gen=6]
Replica r40:/{Table/44-Max} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r40:/{Table/44-Max} [(n2,s2):5, next=6, gen=6]
Conflicting intent found on /Local/Range/System/tsd/RangeDescriptor. Aborting txn d9a0cd59-cbd8-40bb-b56e-444247b74ef7 to resolve.
Scanning replicas on store cluster_id:e90ec120-cc6c-4deb-8573-bd4141b2cf06 node_id:2 store_id:2  for dead peers []
E210726 16:15:14.982494 599845 kv/kvserver/replica.go:894  [n2,s2,r7/5:/Table/1{1-2}] 44  leaseholder replica not in descriptor; desc: r8:/Table/1{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.047986657,0
E210726 16:15:14.985241 599390 kv/kvserver/replica.go:894  [n2,s2,r9/5:/Table/1{3-4}] 45  leaseholder replica not in descriptor; desc: r8:/Table/1{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.047986657,0
E210726 16:15:14.985383 599390 kv/kvserver/replica.go:894  [n2,s2,r9/5:/Table/1{3-4}] 46  leaseholder replica not in descriptor; desc: r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.037571859,0
E210726 16:15:15.011966 600266 kv/kvserver/replica.go:894  [n2,s2,r18/5:/Table/2{2-3}] 47  leaseholder replica not in descriptor; desc: r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.043474978,0
E210726 16:15:15.012108 600266 kv/kvserver/replica.go:894  [n2,s2,r18/5:/Table/2{2-3}] 48  leaseholder replica not in descriptor; desc: r19:/Table/2{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.063974375,0
E210726 16:15:15.070642 601013 kv/kvserver/replica.go:894  [n2,s2,r26/5:/NamespaceTable/{30-Max}] 49  leaseholder replica not in descriptor; desc: r25:/{Table/29-NamespaceTable/30} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.079121679,0
E210726 16:15:15.071333 601013 kv/kvserver/replica.go:894  [n2,s2,r26/5:/NamespaceTable/{30-Max}] 50  leaseholder replica not in descriptor; desc: r27:/{NamespaceTable/Max-Table/32} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.004233084,0
E210726 16:15:15.073424 601026 kv/kvserver/replica.go:894  [n2,s2,r35/5:/Table/{39-40}] 51  leaseholder replica not in descriptor; desc: r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.103036057,0
E210726 16:15:15.073571 601026 kv/kvserver/replica.go:894  [n2,s2,r35/5:/Table/{39-40}] 52  leaseholder replica not in descriptor; desc: r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.081224687,0
E210726 16:15:15.081817 600958 kv/kvserver/replica.go:894  [n2,s2,r6/6:/Table/{SystemCon…-11}] 53  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):6, next=7, gen=10], lease: repl=(n4,s4):2 seq=2 start=1627316094.343992486,0 epo=1 pro=1627316094.343870406,0
E210726 16:15:15.101618 601267 kv/kvserver/replica.go:894  [n2,s2,r27/5:/{NamespaceTab…-Table/32}] 54  leaseholder replica not in descriptor; desc: r28:/Table/3{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.070539108,0
E210726 16:15:15.102330 601266 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 55  leaseholder replica not in descriptor; desc: r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.076508855,0
E210726 16:15:15.124015 601371 kv/kvserver/replica.go:894  [n2,s2,r11/5:/Table/1{5-6}] 56  leaseholder replica not in descriptor; desc: r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.037571859,0
E210726 16:15:15.172925 601561 kv/kvserver/replica.go:894  [n2,s2,r4/6:/System{/tsd-tse}] 57  leaseholder replica not in descriptor; desc: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):6, next=7, gen=10], lease: repl=(n4,s4):2 seq=2 start=1627316094.343992486,0 epo=1 pro=1627316094.343870406,0
E210726 16:15:15.314274 602859 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 58  leaseholder replica not in descriptor; desc: r15:/Table/{19-20} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.100051234,0
E210726 16:15:15.314589 602859 kv/kvserver/replica.go:894  [n2,s2,r16/5:/Table/2{0-1}] 59  leaseholder replica not in descriptor; desc: r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.043474978,0
E210726 16:15:19.715421 599271 kv/kvserver/replica_range_lease.go:365  [n1,s1,r3/1:/System/{NodeLive…-tsd}] 60  failed to heartbeat own liveness record: heartbeat failed on epoch increment

id,is_live,replicas,is_decommissioning,membership,is_draining
3,false,40,true,decommissioning,false
4,false,40,true,decommissioning,false
E210726 16:15:20.280710 635533 kv/kvserver/replica.go:894  [n2,s2,r33/5:/Table/3{7-8}] 61  leaseholder replica not in descriptor; desc: r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.103036057,0
E210726 16:15:20.380843 635532 kv/kvserver/replica.go:894  [n2,s2,r33/5:/Table/3{7-8}] 62  leaseholder replica not in descriptor; desc: r32:/Table/3{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n4,s4):3 seq=2 start=1627316098.048600118,0 epo=1 pro=1627316098.048491743,0
E210726 16:15:20.889018 355188 kv/kvserver/replica.go:894  [n2,s2,r31/5:/Table/3{5-6}] 63  leaseholder replica not in descriptor; desc: r32:/Table/3{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n4,s4):3 seq=2 start=1627316098.048600118,0 epo=1 pro=1627316098.048491743,0
E210726 16:15:21.118301 650615 kv/kvserver/replica.go:894  [n2,s2,r20/5:/Table/2{4-5}] 64  leaseholder replica not in descriptor; desc: r19:/Table/2{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.063974375,0
E210726 16:15:21.118469 650615 kv/kvserver/replica.go:894  [n2,s2,r20/5:/Table/2{4-5}] 65  leaseholder replica not in descriptor; desc: r21:/Table/2{5-6} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316090.989456058,0
E210726 16:15:21.383921 656036 kv/kvserver/replica.go:894  [n2,s2,r28/5:/Table/3{2-3}] 66  leaseholder replica not in descriptor; desc: r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.090260932,0
E210726 16:15:21.919962 654879 kv/kvserver/replica.go:894  [n2,s2,r37/5:/Table/4{1-2}] 67  leaseholder replica not in descriptor; desc: r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.081224687,0
E210726 16:15:21.922658 654879 kv/kvserver/replica.go:894  [n2,s2,r37/5:/Table/4{1-2}] 68  leaseholder replica not in descriptor; desc: r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.056957428,0
E210726 16:15:22.240681 671440 kv/kvserver/replica.go:894  [n2,s2,r40/5:/{Table/44-Max}] 69  leaseholder replica not in descriptor; desc: r39:/Table/4{3-4} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316090.944104805,0
E210726 16:15:22.456902 675504 kv/kvserver/replica.go:894  [n2,s2,r23/5:/Table/2{7-8}] 70  leaseholder replica not in descriptor; desc: r22:/Table/2{6-7} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.073050644,0
E210726 16:15:22.457590 675504 kv/kvserver/replica.go:894  [n2,s2,r23/5:/Table/2{7-8}] 71  leaseholder replica not in descriptor; desc: r24:/Table/2{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.105540854,0
E210726 16:15:23.849027 355188 kv/kvserver/replica.go:894  [n2,s2,r11/5:/Table/1{5-6}] 72  leaseholder replica not in descriptor; desc: r12:/Table/1{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627316099.226244822,0 epo=1 pro=1627316099.226100868,0
E210726 16:15:24.885160 725342 kv/kvserver/replica.go:894  [n2,s2,r22/5:/Table/2{6-7}] 73  leaseholder replica not in descriptor; desc: r21:/Table/2{5-6} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316090.989456058,0
E210726 16:15:25.881410 737751 kv/kvserver/replica.go:894  [n2,s2,r29/5:/Table/3{3-4}] 74  leaseholder replica not in descriptor; desc: r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.076508855,0
E210726 16:15:26.158740 742780 kv/kvserver/replica.go:894  [n2,s2,r15/5:/Table/{19-20}] 75  leaseholder replica not in descriptor; desc: r14:/Table/1{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.087872460,0
E210726 16:15:27.068015 760347 kv/kvserver/replica.go:894  [n2,s2,r13/5:/Table/1{7-8}] 76  leaseholder replica not in descriptor; desc: r14:/Table/1{8-9} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.087872460,0
E210726 16:15:27.256062 355188 kv/kvserver/replica.go:894  [n2,s2,r13/5:/Table/1{7-8}] 77  leaseholder replica not in descriptor; desc: r12:/Table/1{6-7} [(n2,s2):6, next=7, gen=10], lease: repl=(n3,s3):2 seq=2 start=1627316099.226244822,0 epo=1 pro=1627316099.226100868,0
E210726 16:15:27.540879 770156 kv/kvserver/replica.go:894  [n2,s2,r39/5:/Table/4{3-4}] 78  leaseholder replica not in descriptor; desc: r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.056957428,0
E210726 16:15:31.070807 839245 kv/kvserver/replica.go:894  [n2,s2,r24/5:/Table/2{8-9}] 79  leaseholder replica not in descriptor; desc: r25:/{Table/29-NamespaceTable/30} [(n2,s2):5, next=6, gen=6], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1627316091.079121679,0
    debug_test.go:399: expected replicas on {1,2,5,6} but got {1,2,6,7}
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (42.03s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRemoveDeadReplicas PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/server @cockroachdb/kv bdarnell
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cli/debug_test.go
+++ b/pkg/cli/debug_test.go
@@ -132,7 +132,6 @@ func TestOpenReadOnlyStore(t *testing.T) {
 
 func TestRemoveDeadReplicas(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 50977, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	// This test is pretty slow under race (200+ cpu-seconds) because it

--- a/pkg/kv/kvserver/liveness/liveness.go
+++ b/pkg/kv/kvserver/liveness/liveness.go
@@ -76,6 +76,22 @@ func (e *errRetryLiveness) Error() string {
 	return fmt.Sprintf("%T: %s", *e, e.error)
 }
 
+func isErrRetryLiveness(ctx context.Context, err error) bool {
+	if errors.HasType(err, (*roachpb.AmbiguousResultError)(nil)) {
+		// We generally want to retry ambiguous errors immediately, except if the
+		// ctx is canceled - in which case the ambiguous error is probably caused
+		// by the cancellation (and in any case it's pointless to retry with a
+		// canceled ctx).
+		return ctx.Err() == nil
+	} else if errors.HasType(err, (*roachpb.TransactionStatusError)(nil)) {
+		// 21.2 nodes can return a TransactionStatusError when they should have
+		// returned an AmbiguousResultError.
+		// TODO(andrei): Remove this in 22.2.
+		return true
+	}
+	return false
+}
+
 // Node liveness metrics counter names.
 var (
 	metaLiveNodes = metric.Metadata{
@@ -527,41 +543,50 @@ type livenessUpdate struct {
 // NB: An existing liveness record is not overwritten by this method, we return
 // an error instead.
 func (nl *NodeLiveness) CreateLivenessRecord(ctx context.Context, nodeID roachpb.NodeID) error {
-	// We start off at epoch=0, entrusting the initial heartbeat to increment it
-	// to epoch=1 to signal the very first time the node is up and running.
-	liveness := livenesspb.Liveness{NodeID: nodeID, Epoch: 0}
+	for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
+		// We start off at epoch=0, entrusting the initial heartbeat to increment it
+		// to epoch=1 to signal the very first time the node is up and running.
+		liveness := livenesspb.Liveness{NodeID: nodeID, Epoch: 0}
 
-	// We skip adding an expiration, we only really care about the liveness
-	// record existing within KV.
+		// We skip adding an expiration, we only really care about the liveness
+		// record existing within KV.
 
-	v := new(roachpb.Value)
-	if err := nl.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-		b := txn.NewBatch()
-		key := keys.NodeLivenessKey(nodeID)
-		if err := v.SetProto(&liveness); err != nil {
-			log.Fatalf(ctx, "failed to marshall proto: %s", err)
-		}
-		// Given we're looking to create a new liveness record here, we don't
-		// expect to find anything.
-		b.CPut(key, v, nil)
+		v := new(roachpb.Value)
+		err := nl.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			b := txn.NewBatch()
+			key := keys.NodeLivenessKey(nodeID)
+			if err := v.SetProto(&liveness); err != nil {
+				log.Fatalf(ctx, "failed to marshall proto: %s", err)
+			}
+			// Given we're looking to create a new liveness record here, we don't
+			// expect to find anything.
+			b.CPut(key, v, nil)
 
-		// We don't bother adding a gossip trigger, that'll happen with the
-		// first heartbeat. We still keep it as a 1PC commit to avoid leaving
-		// write intents.
-		b.AddRawRequest(&roachpb.EndTxnRequest{
-			Commit:     true,
-			Require1PC: true,
+			// We don't bother adding a gossip trigger, that'll happen with the
+			// first heartbeat. We still keep it as a 1PC commit to avoid leaving
+			// write intents.
+			b.AddRawRequest(&roachpb.EndTxnRequest{
+				Commit:     true,
+				Require1PC: true,
+			})
+			return txn.Run(ctx, b)
 		})
-		return txn.Run(ctx, b)
-	}); err != nil {
+
+		if err == nil {
+			// We'll learn about this liveness record through gossip eventually, so we
+			// don't bother updating our in-memory view of node liveness.
+			log.Infof(ctx, "created liveness record for n%d", nodeID)
+			return nil
+		}
+		if !isErrRetryLiveness(ctx, err) {
+			return err
+		}
+		log.VEventf(ctx, 2, "failed to create liveness record for node %d, because of %s. retrying...", nodeID, err)
+	}
+	if err := ctx.Err(); err != nil {
 		return err
 	}
-
-	// We'll learn about this liveness record through gossip eventually, so we
-	// don't bother updating our in-memory view of node liveness.
-
-	log.Infof(ctx, "created liveness record for n%d", nodeID)
-	return nil
+	return errors.AssertionFailedf("unexpected problem while creating liveness record for node %d", nodeID)
 }
 
 func (nl *NodeLiveness) setMembershipStatusInternal(
@@ -1318,19 +1343,7 @@ func (nl *NodeLiveness) updateLivenessAttempt(
 				return Record{}, errors.Wrapf(err, "couldn't update node liveness from CPut actual value")
 			}
 			return Record{}, handleCondFailed(Record{Liveness: actualLiveness, raw: tErr.ActualValue.TagAndDataBytes()})
-		} else if errors.HasType(err, (*roachpb.AmbiguousResultError)(nil)) {
-			// We generally want to retry ambiguous errors immediately, except if the
-			// ctx is canceled - in which case the ambiguous error is probably caused
-			// by the cancellation (and in any case it's pointless to retry with a
-			// canceled ctx).
-			if ctx.Err() != nil {
-				return Record{}, err
-			}
-			return Record{}, &errRetryLiveness{err}
-		} else if errors.HasType(err, (*roachpb.TransactionStatusError)(nil)) {
-			// 21.2 nodes can return a TransactionStatusError when they should have
-			// returned an AmbiguousResultError.
-			// TODO(andrei): Remove this in 22.2.
+		} else if isErrRetryLiveness(ctx, err) {
 			return Record{}, &errRetryLiveness{err}
 		}
 		return Record{}, err

--- a/pkg/kv/kvserver/node_liveness_test.go
+++ b/pkg/kv/kvserver/node_liveness_test.go
@@ -1053,6 +1053,76 @@ func TestNodeLivenessRetryAmbiguousResultError(t *testing.T) {
 	}
 }
 
+// This tests the create code path for node liveness, for that we need to create
+// a cluster of nodes, because we need to exercise to join RPC codepath.
+func TestNodeLivenessRetryAmbiguousResultOnCreateError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	var injectError atomic.Value
+	type injectErrorData struct {
+		shouldError bool
+		count       int32
+	}
+
+	injectError.Store(map[roachpb.NodeID]injectErrorData{
+		roachpb.NodeID(1): {true, 0},
+		roachpb.NodeID(2): {true, 0},
+		roachpb.NodeID(3): {true, 0},
+	})
+	testingEvalFilter := func(args kvserverbase.FilterArgs) *roachpb.Error {
+		if req, ok := args.Req.(*roachpb.ConditionalPutRequest); ok {
+			if val := injectError.Load(); val != nil {
+				var liveness livenesspb.Liveness
+				if err := req.Value.GetProto(&liveness); err != nil {
+					return nil
+				}
+				injectErrorMap := val.(map[roachpb.NodeID]injectErrorData)
+				if injectErrorMap[liveness.NodeID].shouldError {
+					if liveness.NodeID != 1 {
+						// We expect this to come from the create code path on all nodes
+						// except the first. Make sure that is actually true.
+						assert.Equal(t, liveness.Epoch, int64(0))
+					}
+					injectErrorMap[liveness.NodeID] = injectErrorData{false, injectErrorMap[liveness.NodeID].count + 1}
+					injectError.Store(injectErrorMap)
+					return roachpb.NewError(roachpb.NewAmbiguousResultError("test"))
+				}
+			}
+		}
+		return nil
+	}
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 3,
+		base.TestClusterArgs{
+			ReplicationMode: base.ReplicationManual,
+			ServerArgs: base.TestServerArgs{
+				Knobs: base.TestingKnobs{
+					Store: &kvserver.StoreTestingKnobs{
+						EvalKnobs: kvserverbase.BatchEvalTestingKnobs{
+							TestingEvalFilter: testingEvalFilter,
+						},
+					},
+				},
+			},
+		})
+	defer tc.Stopper().Stop(ctx)
+
+	for _, s := range tc.Servers {
+		// Verify retry of the ambiguous result for heartbeat loop.
+		testutils.SucceedsSoon(t, func() error {
+			return verifyLivenessServer(s, 3)
+		})
+		nl := s.NodeLiveness().(*liveness.NodeLiveness)
+		_, ok := nl.Self()
+		assert.True(t, ok)
+
+		injectErrorMap := injectError.Load().(map[roachpb.NodeID]injectErrorData)
+		assert.NotNil(t, injectErrorMap)
+		assert.Equal(t, int32(1), injectErrorMap[s.NodeID()].count)
+	}
+}
+
 // Test that, although a liveness heartbeat is generally retried on
 // AmbiguousResultError (see test above), it is not retried when the error is
 // caused by a canceled context.

